node/engine/chainservice/eth_chainservice.go

```diff
@@ -123,7 +132,12 @@ func newEthChainService(chain ethChain, na *NitroAdjudicator.NitroAdjudicator,
 		return nil, err
 	}
 
-	// TODO: Return error from chain service instead of panicking
+	// Search for any missed events emitted while this node was offline
+	err = ecs.checkForMissedEvents(startBlock)
```

Comment:
I think in theory there's a risk here that a block (or blocks) could get mined while `checkForMissedEvents` is executing, before we start listening for new events in `ecs.listenForEventLogs`. This means the events in those blocks would be "ignored". In practice this is incredibly unlikely since `checkForMissedEvents` is just pushing events into the `event-tracker` and thus is really fast. Blocks are mined at a pretty slow pace in comparison, so it's very unlikely a block will be mined in between `checkForMissedEvents` and when we start the listener.

Maybe it's worth starting the listeners before calling `checkMissedEvents`  just to be on the safe side? Once we know we're listening to any new events we can call `checkMissedEvents` to fetch the missed ones.

Reply:
Yes thats a valid concern. I was thinking about it as part of [this adr](https://github.com/statechannels/go-nitro/blob/b82b0dd1aef34b0fa84cf6ed02b8ef99f46be605/.adr/0017-missed-chain-events.md#future-considerations)). 

My counter-concern if we start `go ecs.listenForEventLogs` before `ecs.checkForMissedEvents` is this scenario (which is probably less likely to occur, but ):

1. Block 100 mined
2. `listenForEventLogs` queues events `AE25`
3. `checkForMissedEvents` starts queuing events `AE1`, `AE2`, etc.
4. Block 101 mined
5. `checkForMissedEvents` queues events `AE3`, `AE4`
6. Block 102 mined
7. `dispatchChainEvents([AE1, AE2, AE3, AE4, AE25])` is triggered
8. `checkForMissedEvents` queues events `AE5`, `AE6` and completes
9. `dispatchChainEvents([AE5, AE6])` is triggered
10. `AE5` and `AE6` can't be processed if they're related to same channel as `AE25`

This is a very low concern if the time between blocks is long because `checkForMissedEvents` should be fast. But I just wanted to mention it in case there is a scenario like this:

* blockchain with fast block times
* `checkForMissedEvents` takes a long time to complete because it has to search through many blocks (nitro node has been off for a long time)

Reply:
> This is a very low concern if the time between blocks is long because `checkForMissedEvents` should be fast. But I just wanted to mention it in case there is a scenario like this:
> 
>     * blockchain with fast block times
> 
>     * `checkForMissedEvents` takes a long time to complete because it has to search through many blocks (nitro node has been off for a long time)

Unfortunately this is also the exact scenario where we would expect to miss events when calling `checkForMissedEvents` before starting `ecs.listenForEventLog`.

I think the "full solution" may involve listening for events before `checkForMissedEvents` but only dispatching events (in order) after `checkForMissedEvents` completes. That way we're protected against missing events but enforce that events are dispatched in the correct order.

Reply:
Could we use the mutex on the eventTracker as a way to ensure events are NOT dispatched before `checkForMissedEvents` finishes queuing old events? Something like this:

```
	// Prevent go routines from processing events before checkForMissedEvents completes
	ecs.eventTracker.mu.Lock()
	
	ecs.wg.Add(3)
	go ecs.listenForEventLogs(errChan, eventSub, eventChan, eventQuery)
	go ecs.listenForNewBlocks(errChan, newBlockSub, newBlockChan)
	go ecs.listenForErrors(errChan)

	// Search for any missed events emitted while this node was offline
	err = ecs.checkForMissedEvents(startBlock)
	if err != nil {
		return nil, err
	}
	ecs.eventTracker.mu.Unlock()
```

`listenForEventLogs` calls `updateEventTracker`, which also tries to acquire the `mu.Lock()`. That should be blocked until the `mu.Unlock()` is called in the init routine code shown above.

Reply:
Added above implementation in https://github.com/statechannels/go-nitro/pull/1609/commits/edade6ed43a5fce06386c6345afff7e81d7917f5

Reply:
Nice solution 🚀 ! By locking on the mutex we block the listeners from calling `updateEventTracker`, but since the `checkForMissedEvents` doesn't acquire a mutex it will still be able to write to the event tracker.

I think it's beyond the scope of this PR but we may want to consider refactoring the `eventTracker` so it exposes functions (`AddEvent`,`StartDispatchingEvents`) instead of requiring the chain service to deal with the event heap and acquiring mutexes. The chain service is already quite complex, so I think it would be a win to refactor some of the complexity into another struct.

Reply:
That's a good refactor suggestion. I'll address in a follow-up PR